src/main.cpp

```diff
@@ -859,22 +890,51 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         CTxMemPoolEntry entry(tx, nFees, GetTime(), dPriority, chainActive.Height(), mempool.HasNoInputsOf(tx));
         unsigned int nSize = entry.GetTxSize();
 
-        // Don't accept it if it can't get into a block
-        CAmount txMinFee = GetMinRelayFee(tx, nSize, true);
-        if (fLimitFree && nFees < txMinFee)
-            return state.DoS(0, error("AcceptToMemoryPool: not enough fees %s, %d < %d",
-                                      hash.ToString(), nFees, txMinFee),
-                             REJECT_INSUFFICIENTFEE, "insufficient fee");
-
-        // Require that free transactions have sufficient priority to be mined in the next block.
-        if (GetBoolArg("-relaypriority", true) && nFees < ::minRelayTxFee.GetFee(nSize) && !AllowFree(view.GetPriority(tx, chainActive.Height() + 1))) {
-            return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "insufficient priority");
+        // The fees required to accept this transaction start with the fees required to accept it on its own
+        CAmount nFeesRequired = 0;
+        if (fLimitFree) {
+            nFeesRequired = GetMinRelayFee(tx, nSize, true);
+            if (nFees < nFeesRequired)
+                return state.DoS(0, error("AcceptToMemoryPool: not enough fees %s, %d < %d", hash.ToString(), nFees, nFeesRequired),
+                                 REJECT_INSUFFICIENTFEE, "insufficient fee");
+        }
+        // If we are not relaying low priority free transactions, then if this tx doesn't have sufficient priority
+        // it must have minRelayTxFee
+        if (GetBoolArg("-relaypriority", true) && nFeesRequired < ::minRelayTxFee.GetFee(nSize) && !AllowFree(view.GetPriority(tx, chainActive.Height() + 1))) {
+            nFeesRequired =  ::minRelayTxFee.GetFee(nSize);
+            if (nFees < nFeesRequired)
+                return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "insufficient priority");
+        }
+
+        // Try to make space in mempool
+        size_t softcap = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 700000;
+        size_t hardcap = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000;
+        size_t capstep = (hardcap - softcap) / 10;
+        std::set<uint256> stagedelete;
+        CAmount nFeesDeleted = 0;
+
+        if (!mempool.StageTrimToSize(softcap, entry, stagedelete, nFeesRequired, nFeesDeleted)) {
```

Comment:
@sipa   Neither is the answer.  Perhaps it needs commenting though.  If the transaction is accepted into the reserve space and stagedelete is non-empty then it is both safe and desirable to evict any transactions that were staged.   If the transaction is not accepted into the reserve space, then as before we return failure from ATMP before we have a chance to evict anything.

The one slight caveat is in the case where you pass entry into the reserve space, but then you fail the rate limiter code directly below it.  In that case, it probably wasn't very fair that your feesDeleted were causing you to be counted as a free transaction.  It's overly conservative and rare behavior though.  I was planning on suggesting a fix but I wanted to see if there were any thoughts on the free relay code remaining in place here before I complicated it.  The suggested fix is that if the transaction is entering the mempool via the reserve space, and then it fails the rate limiter code, you let it in anyway but just wipe stagedelete.  Note that if stagedelete were empty then the tx by definition wouldn't have been subject to the rate limit test because it has a multiple of the minRelayFee.